### PR TITLE
Fix swsscommon psubscribe code break in frrcfgd

### PR DESCRIPTION
**Pull request by liuh-80**
Fix swsscommon psubscribe code break in frrcfgd

#### Why I did it
Fix frrcfgd psubscribe code break: https://github.com/sonic-net/sonic-buildimage/issues/13109
The code issue caused by API change when migrate from swsssdk to swsscommon

#### How I did it
Fix frrcfgd code to use swsscommon psubscribe API.

#### How to verify it
Pass all UT.
Manually check fixed code work correctly.

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [x] 202211

#### Description for the changelog
Fix frrcfgd psubscribe code break: https://github.com/sonic-net/sonic-buildimage/issues/13109

### CODE REVIEW

File: src/sonic-frr-mgmt-framework/frrcfgd/frrcfgd.py [hunk 1]
```diff
@@ -1445,6 +1445,7 @@ class ExtConfigDBConnector(ConfigDBConnector):
     def __init__(self, ns_attrs = None):
         super(ExtConfigDBConnector, self).__init__()
         self.nosort_attrs = ns_attrs if ns_attrs is not None else {}
+        self.sub_thread_running = False
```
Comment (qiluo-msft) [inline src/sonic-frr-mgmt-framework/frrcfgd/frrcfgd.py:1448]: >sub_thread_running [](http://example.com/codeflow?start=13&length=18)

The member is expected to be private, follow python convention, use prefix `_` or `__` to indicate the intention.

File: src/sonic-frr-mgmt-framework/frrcfgd/frrcfgd.py [hunk 2]
```diff
@@ -1469,12 +1470,28 @@ def sub_msg_handler(self, msg_item):
...
+    def listen_thread(self, timeout):
+        self.sub_thread_running = True
+        sub_key_space = "__keyspace@{}__:*".format(self.get_dbid(self.db_name))
+        self.pubsub.psubscribe(sub_key_space)
+        while self.sub_thread_running:
+            msg = self.pubsub.get_message(timeout, True)
+            if msg:
+                self.sub_msg_handler(msg)
+
+        self.pubsub.punsubscribe(sub_key_space)
```
Comment (qiluo-msft) [inline src/sonic-frr-mgmt-framework/frrcfgd/frrcfgd.py:1483]: >punsubscribe [](http://example.com/codeflow?start=20&length=12)

Found similar code in [https://github.com/sonic-net/sonic-swss-common/blob/85f3776ae6da44a083c617ff4ce6e54e1e2d5dac/common/configdb.h#L78](https://github.com/sonic-net/sonic-swss-common/blob/85f3776ae6da44a083c617ff4ce6e54e1e2d5dac/common/configdb.h#L78). Do we also need `punsubscribe` there?

**Comment by qiluo-msft** [comment]
@zhaozhenhong Could you help review?

File: src/sonic-frr-mgmt-framework/frrcfgd/frrcfgd.py (hunk 2 above)
Comment (liuh-80) [reply src/sonic-frr-mgmt-framework/frrcfgd/frrcfgd.py:1483]: Yes, I think we also unsubscribe after exit the while loop in ConfigDBConnector, I will create PR for this.

File: src/sonic-frr-mgmt-framework/frrcfgd/frrcfgd.py (hunk 1 above)
Comment (liuh-80) [reply src/sonic-frr-mgmt-framework/frrcfgd/frrcfgd.py]: Fixed

**Comment by jeff-yin** [comment]
Please also tag @mhampasagar and @rathnasabapathyv  for review, thanks.

**Review by mhampasagar — APPROVED** [review]

**Review by qiluo-msft — APPROVED** [review]

**Comment by puffc** [comment]
@qiluo-msft Would you please merge this PR to 202211? Thanks!

**Comment by bluecmd** [comment]
This change caused my frrcfgd to spike to 100% CPU usage on my devices. I changed the default timeout to 2 seconds instead of 1 ms and that seemed to solve things.

```patch
diff --git a/src/sonic-frr-mgmt-framework/frrcfgd/frrcfgd.py b/src/sonic-frr-mgmt-framework/frrcfgd/frrcfgd.py
index bc53e38cf..6e53311c7 100755
--- a/src/sonic-frr-mgmt-framework/frrcfgd/frrcfgd.py
+++ b/src/sonic-frr-mgmt-framework/frrcfgd/frrcfgd.py
@@ -1471,12 +1471,12 @@ class ExtConfigDBConnector(ConfigDBConnector):
                 syslog.syslog(syslog.LOG_ERR, '[bgp cfgd] Failed handling config DB update with exception:' + str(e))
                 logging.exception(e)
 
-    def listen_thread(self, timeout):
+    def listen_thread(self):
         self.__listen_thread_running = True
         sub_key_space = "__keyspace@{}__:*".format(self.get_dbid(self.db_name))
         self.pubsub.psubscribe(sub_key_space)
         while self.__listen_thread_running:
-            msg = self.pubsub.get_message(timeout, True)
+            msg = self.pubsub.get_message(timeout=2.0, ignore_subscribe_messages=True)
             if msg:
                 self.sub_msg_handler(msg)
 
@@ -1486,7 +1486,7 @@ class ExtConfigDBConnector(ConfigDBConnector):
         """Start listen Redis keyspace events and will trigger corresponding handlers when content of a table changes.
         """
         self.pubsub = self.get_redis_client(self.db_name).pubsub()
-        self.sub_thread = threading.Thread(target=self.listen_thread, args=(0.01,))
+        self.sub_thread = threading.Thread(target=self.listen_thread)
         self.sub_thread.start()
 
     def stop_listen(self):
```

**Comment by liuh-80** [comment]
> This change caused my frrcfgd to spike to 100% CPU usage on my devices. I changed the default timeout to 2 seconds instead of 1 ms and that seemed to solve things.
> 
> ```diff
> diff --git a/src/sonic-frr-mgmt-framework/frrcfgd/frrcfgd.py b/src/sonic-frr-mgmt-framework/frrcfgd/frrcfgd.py
> index bc53e38cf..6c18008f1 100755
> --- a/src/sonic-frr-mgmt-framework/frrcfgd/frrcfgd.py
> +++ b/src/sonic-frr-mgmt-framework/frrcfgd/frrcfgd.py
> @@ -1471,12 +1471,12 @@ class ExtConfigDBConnector(ConfigDBConnector):
>                  syslog.syslog(syslog.LOG_ERR, '[bgp cfgd] Failed handling config DB update with exception:' + str(e))
>                  logging.exception(e)
>  
> -    def listen_thread(self, timeout):
> +    def listen_thread(self):
>          self.__listen_thread_running = True
>          sub_key_space = "__keyspace@{}__:*".format(self.get_dbid(self.db_name))
>          self.pubsub.psubscribe(sub_key_space)
>          while self.__listen_thread_running:
> -            msg = self.pubsub.get_message(timeout, True)
> +            msg = self.pubsub.get_message(timeout=2.0, interrupt_on_signal=True)
>              if msg:
>                  self.sub_msg_handler(msg)
>  
> @@ -1486,7 +1486,7 @@ class ExtConfigDBConnector(ConfigDBConnector):
>          """Start listen Redis keyspace events and will trigger corresponding handlers when content of a table changes.
>          """
>          self.pubsub = self.get_redis_client(self.db_name).pubsub()
> -        self.sub_thread = threading.Thread(target=self.listen_thread, args=(0.01,))
> +        self.sub_thread = threading.Thread(target=self.listen_thread)
>          self.sub_thread.start()
>  
>      def stop_listen(self):
> ```

Hi @bluecmd, can you share how to reproduce this issue and what's the hardware SKU and OS version?

**Comment by bluecmd** [comment]
Hello @liuh-80 

Sure - I applied this PR on our 202211 branch and installed the resulting image on both a Celestica Questone 2 and a Celestica Silverstone-X. When running I looked at `top` and saw that frrcfgd was almost always at 100% CPU.

**Comment by liuh-80** [comment]
> Hello @liuh-80
> 
> Sure - I applied this PR on our 202211 branch and installed the resulting image on both a Celestica Questone 2 and a Celestica Silverstone-X. When running I looked at `top` and saw that frrcfgd was almost always at 100% CPU.

Hi @bluecmd , I check on following celestica hardware and not found this issue, I will create some test code and give update later.

SONiC Software Version: SONiC.20230531.19
SONiC OS Version: 11
Distribution: Debian 11.8
Kernel: 5.10.0-23-2-amd64
Build commit: 2504c06bcc
Build date: Sun Feb 25 03:18:23 UTC 2024
Built by: cloudtest@361280d0c000000

Platform: x86_64-cel_e1031-r0
HwSKU: Celestica-E1031-T48S4
ASIC: broadcom
ASIC Count: 1
Serial Number: R0882F2B128709BY000006
Model Number: R0882-F4111-01
Hardware Revision: N/A
Uptime: 07:21:38 up 11:34,  1 user,  load average: 5.45, 7.10, 5.57
Date: Thu 29 Feb 2024 07:21:38



~$ top
top - 07:21:28 up 11:34,  1 user,  load average: 5.53, 7.18, 5.57
Tasks: 275 total,   8 running, 266 sleeping,   0 stopped,   1 zombie
%Cpu(s): 78.3 us, 21.3 sy,  0.0 ni,  0.0 id,  0.0 wa,  0.0 hi,  0.5 si,  0.0 st
MiB Mem :   1948.3 total,     91.3 free,   1327.9 used,    529.0 buff/cache
MiB Swap:   2048.0 total,   1253.5 free,    794.5 used.    430.9 avail Mem

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
 917654 root      20   0   65316  41512  17584 R  33.1   2.1   0:01.03 show
 910644 root      20   0 1869700 162264  23540 S  19.9   8.1   1:10.52 syncd
 917684 root      20   0   29200  22260   9176 R  17.4   1.1   0:00.54 sonic-cfggen
 917616 root      20   0   71844  44428  20000 S  13.5   2.2   0:01.14 python3

**Comment by bluecmd** [comment]
Interesting. Let me know if you need me to test anything. 

That said, if I may suggest, I believe that the 0.01s timeout is too low even on platforms where it does not show up as high CPU - since the subscribe is running in its own thread it only needs to wake up occasionally to check if the thread should terminate, and that does not need to happen often at all. I have not checked, but my assumption is that creating and destroying PubSub subscriptions incur a cost, so we should probably not do it needlessly. That's my analysis anyway.

**Comment by liuh-80** [comment]
> Interesting. Let me know if you need me to test anything.
> 
> That said, if I may suggest, I believe that the 0.01s timeout is too low even on platforms where it does not show up as high CPU - since the subscribe is running in its own thread it only needs to wake up occasionally to check if the thread should terminate, and that does not need to happen often at all. ~I have not checked, but my assumption is that creating and destroying PubSub subscriptions incur a cost, so we should probably not do it needlessly. That's my analysis anyway.~ EDIT: The subscribe is outside the loop, ignore the last part

I can reproduce this issue by a test script, fix PR is here: https://github.com/sonic-net/sonic-buildimage/pull/18225

**Pull request merged**